packages/muelu/src/Transfers/GeneralGeometric/MueLu_GeneralGeometricPFactory_def.hpp

```diff
@@ -0,0 +1,843 @@
+  LocalOrdinal GeneralGeometricPFactory<Scalar, LocalOrdinal, GlobalOrdinal, Node>::MakeGeneralGeometricP(LocalOrdinal const ndm, LocalOrdinal const nFpts[], LocalOrdinal const nCpts[], LocalOrdinal const coarse_rate[], LocalOrdinal const end_rate[], const RCP<Xpetra::MultiVector<double,LocalOrdinal,GlobalOrdinal,Node> > fCoords, LocalOrdinal const nnzP, LocalOrdinal const dofsPerNode, RCP<const Map>& coarseMap, RCP<Matrix> & Amat, RCP<Matrix>& P, RCP<Xpetra::MultiVector<double,LocalOrdinal,GlobalOrdinal,Node> >& cCoords) const {
+
+    /*
+     * Given a CSR matrix (OrigARowPtr, OrigAcols, OrigAvals), information
+     * describing the z-layer and vertical line (LayerId and VertLineId)
+     * of each matrix block row, a coarsening rate, and dofs/node information,
+     * construct a prolongator that coarsening to semicoarsening in the z-direction
+     * using something like an operator-dependent grid transfer. In particular,
+     * matrix stencils are collapsed to vertical lines. Thus, each vertical line
+     * gives rise to a block tridiagonal matrix. BlkRows corresponding to
+     * Cpts are replaced by identity matrices. This tridiagonal is solved
+     * to determine each interpolation basis functions. Each Blk Rhs corresponds
+     * to all zeros except at the corresponding C-pt which has an identity
+     *
+     * On termination, return the number of local prolongator columns owned by
+     * this processor.
+     *
+     * Input
+     * =====
+     *    nNodes       Number of fine level Blk Rows owned by this processor
+     *    coarse_rate  Rate of coarsening in each spatial direction.
+     *    end_rate     Rate of coarsening in each spatial direction for the last
+     *                 nodes in the mesh where an adaptive coarsening rate is 
+     *                 required.
+     *    nTerms       Number of nonzero entries in the prolongation matrix.
+     *    dofsPerNode  Number of degrees-of-freedom per mesh node.
+     *
+     * Output
+     * =====
+     *    So far nothing...
+     */
+    
+    typedef Teuchos::ScalarTraits<SC> STS;
+    const SC     zero      = STS::zero();
+    const SC     one       = STS::one();
+    const LO     INVALID   = Teuchos::OrdinalTraits<LO>::invalid();
+    
+    // RCP<Teuchos::FancyOStream> fancy = Teuchos::fancyOStream(Teuchos::rcpFromRef(std::cout));
+    // Teuchos::FancyOStream& out = *fancy;
+
+    LO nFineNodes = nFpts[0]*nFpts[1]*nFpts[2];
+    LO numRows = dofsPerNode*nFineNodes;
+    LO nCoarseNodes = nCpts[0]*nCpts[1]*nCpts[2];
+    LO numGloCols=dofsPerNode*nCoarseNodes;
+    LO numLocCols= dofsPerNode*nCoarseNodes; // This Needs to change for parallel implementation!
+    LO maxEntriesPerRow=0;
+    if(ndm==1) {
+      maxEntriesPerRow=2;
+    } else if(ndm==2) {
+      maxEntriesPerRow=4;
+    } else if(ndm==3) {
+      maxEntriesPerRow=8;
+    }
+
+    RCP<const Map> rowMap = Amat->getRowMap();
+    std::vector<size_t> stridingInfo_;
```

Comment:
Because I have seen it in another transfer operator factory and thought this might be the standard?